library/signed/src/Operations.qs

```diff
@@ -0,0 +1,356 @@
+                within {
+                    AND(controls[0], actl, aux);
+                } apply {
+                    Controlled AddI([aux], (ys, (result[idx..idx + nb])));
+                }
+            }
+        } else {
+            use helper = Qubit[numControls];
+            within {
+                AndLadder(controls, Most(helper));
+            } apply {
+                for (idx, actl) in Enumerated(xs) {
+                    within {
+                        AND(Tail(Most(helper)), actl, Tail(helper));
+                    } apply {
+                        Controlled AddI([Tail(helper)], (ys, (result[idx..idx + nb])));
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// # Summary
+/// Multiply signed integer `xs` by signed integer `ys` and store
+/// the result in `result`, which must be zero initially.
+///
+/// # Input
+/// ## xs
+/// 𝑛₁-bit multiplicand
+/// ## ys
+/// 𝑛₂-bit multiplier
+/// ## result
+/// (𝑛₁+𝑛₂)-bit result, must be in state |0⟩
+/// initially.
+operation MultiplySI(xs : Qubit[], ys : Qubit[], result : Qubit[]) : Unit is Adj + Ctl {
+    body (...) {
+        Controlled MultiplySI([], (xs, ys, result));
+    }
+    controlled (controls, ...) {
+        use signx = Qubit();
+        use signy = Qubit();
+
+        within {
+            CNOT(Tail(xs), signx);
+            CNOT(Tail(ys), signy);
+            Controlled Invert2sSI([signx], xs);
+            Controlled Invert2sSI([signy], ys);
+        } apply {
+            Controlled MultiplyI(controls, (xs, ys, result));
+            within {
+                CNOT(signx, signy);
+            } apply {
+                // No controls required since `result` will still be zero
+                // if we did not perform the multiplication above.
+                Controlled Invert2sSI([signy], result);
+            }
+        }
+    }
+}
+
+
+
+/// # Summary
+/// Implements a reversible sum gate. Given a carry-in bit encoded in
+/// qubit `carryIn` and two summand bits encoded in `summand1` and `summand2`,
+/// computes the bitwise xor of `carryIn`, `summand1` and `summand2` in the qubit
+/// `summand2`.
+///
+/// # Input
+/// ## carryIn
+/// Carry-in qubit.
+/// ## summand1
+/// First summand qubit.
+/// ## summand2
+/// Second summand qubit, is replaced with the lower bit of the sum of
+/// `summand1` and `summand2`.
+///
+/// # Remarks
+/// In contrast to the `Carry` operation, this does not compute the carry-out bit.
+operation Sum(carryIn : Qubit, summand1 : Qubit, summand2 : Qubit) : Unit is Adj + Ctl {
+    CNOT(summand1, summand2);
+    CNOT(carryIn, summand2);
+}
+
+
+/// # Summary
+/// Automatically chooses between addition with
+/// carry and without, depending on the register size of `ys`.
+///
+/// # Input
+/// ## xs
+/// $n$-bit addend.
+/// ## ys
+/// Addend with at least $n$ qubits. Will hold the result.
+operation AddI(xs : Qubit[], ys : Qubit[]) : Unit is Adj + Ctl {
```

Comment:
I suspect you could use RippleCarryTTKIncByLE instead of this entire operation.

Reply:
The tests still pass with this change, so I'll do it! Thanks.